data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/export/DataFileLoader.java

```diff
@@ -0,0 +1,182 @@
+            this.bucketName = bucketName;
+            return this;
+        }
+
+        public Builder key(String key) {
+            this.key = key;
+            return this;
+        }
+
+        public Builder startLine(int startLine) {
+            this.startLine = startLine;
+            return this;
+        }
+
+        public DataFileLoader build() {
+            return new DataFileLoader(this);
+        }
+
+    }
+
+
+    @Override
+    public void run() {
+        LOG.debug("Read export data from s3://" + bucketName + "/" + key + " with start line " + startLine);
+        long lastCheckpointTime = System.currentTimeMillis();
+        List<String> lines = new ArrayList<>();
+
+        // line count regardless the start line number
+        int lineCount = 0;
+        int lastLineProcessed = 0;
+
+        try (GZIPInputStream gzipInputStream = new GZIPInputStream(s3ObjectReader.readFile(bucketName, key))) {
+            BufferedReader reader = new BufferedReader(new InputStreamReader(gzipInputStream));
+
+            String line;
+            while ((line = reader.readLine()) != null) {
+
+                if (shouldStop) {
+                    checkpointer.checkpoint(lastLineProcessed);
+                    LOG.debug("Should Stop flag is set to True, looks like shutdown has triggered");
+                    throw new RuntimeException("Load is interrupted");
+                }
+
+                lineCount += 1;
+
+                // process each line
+                if (lineCount <= startLine) {
+                    continue;
+                }
+                lines.add(line);
+
+                if ((lineCount - startLine) % DEFAULT_BATCH_SIZE == 0) {
+                    // LOG.debug("Write to buffer for line " + (lineCount - DEFAULT_BATCH_SIZE) + " to " + lineCount);
+                    recordConverter.writeToBuffer(lines);
+                    lines.clear();
+                    lastLineProcessed = lineCount;
+                }
+
+                if (System.currentTimeMillis() - lastCheckpointTime > DEFAULT_CHECKPOINT_INTERVAL_MILLS) {
+                    LOG.debug("Perform regular checkpointing for Data File Loader");
+                    checkpointer.checkpoint(lastLineProcessed);
+                    lastCheckpointTime = System.currentTimeMillis();
+
+                }
+
+            }
+            if (!lines.isEmpty()) {
+                // Do final checkpoint.
+                recordConverter.writeToBuffer(lines);
+                checkpointer.checkpoint(lineCount);
+            }
+
+            lines.clear();
+            reader.close();
+            LOG.debug("Data Loader completed successfully");
+        } catch (Exception e) {
+            checkpointer.checkpoint(lineCount);
+            throw new RuntimeException("Data Loader completed with Exception: " + e.getMessage());
+        }
+
+    }
+
+    /**
+     * Currently, this is to stop all consumers.
+     */
+    public static void stopAll() {
```

Comment:
Since this is static, it will stop all consumers in all pipelines. Use an instance per pipeline instead.

Reply:
Why this has impacts to other pipelines, if they are running in separate containers or different nodes?

Reply:
Data Prepper can run multiple parallel pipelines. If you have two pipelines with the `dynamodb` source they will share the same static `shouldStop` variable. Thus, when one pipeline shuts down, the other pipeline's DataFileLoader is effectively stopped. This may not be what the user wants.